routes/controllers/users.js

```diff
@@ -0,0 +1,187 @@
+import jwt from "jsonwebtoken";
+import bCrypt from "bcryptjs";
+import service from "../../services/users.js";
+
+import {
+  userRegisterSchema,
+  userLoginSchema,
+  userLogoutSchema,
+  userSubSchema,
+} from "../../utils/validation.js";
+
+const register = async (req, res, next) => {
+  try {
+    const { body } = req;
+    const { username, email, password } = body;
+    await userRegisterSchema.validateAsync(body);
+
+    const isUserExists = await service.getUserWithOrOperator([
+      { username },
+      { email },
+    ]);
+
+    if (isUserExists) {
+      return res.status(409).json({
+        status: 409,
+        statusText: "Conflict",
+        data: {
+          message: `${
+            isUserExists.username === username
+              ? "Username"
+              : isUserExists.email === email
+              ? "E-mail"
+              : null
+          } is already in use`,
+        },
+      });
+    }
+
+    const user = await service.createUser(body);
+    await user.validate();
+    user.set(
+      "password",
+      await bCrypt.hash(password, await bCrypt.genSalt(6)),
+      String
+    );
+    await user.save();
+
+    res.status(201).json({
+      status: 201,
+      statusText: "Created",
+      data: {
+        user: {
+          email: user.email,
+          subscription: user.subscription,
+        },
+      },
+    });
+  } catch (err) {
+    res.status(400).json({
+      status: 400,
+      statusText: "Bad Request",
+      data: { message: err.message },
+    });
+  }
+};
+
+const login = async (req, res, next) => {
+  try {
+    await userLoginSchema.validateAsync(req.body);
+    const { email, password } = req.body;
+    const existingUser = await service.getUser({ email });
+
+    if (
+      !existingUser ||
+      !(await bCrypt.compare(password, existingUser.password))
+    ) {
+      return res.status(401).json({
+        status: 401,
+        statusText: "Unauthorized",
+        data: { message: "Incorrect e-mail or password" },
+      });
+    }
+
+    const payload = {
+      id: existingUser._id,
+    };
+   
+
+    const token = jwt.sign(payload, process.env.JWT_SECRET, {
+      expiresIn: "1h",
+    });
+
+
+      
+    const user = await service.updateUser({ email }, { token });
+
+    res.json({
+      status: 200,
+      statusText: "OK",
+      data: {
+        token,
+        user: {
+          email: user.email,
+          subscription: user.subscription,
+        },
+      },
+    });
+  } catch (err) {
+    res.status(400).json({
+      status: 400,
+      statusText: "Bad Request",
+      data: { message: err.message },
+    });
+  }
+};
+
+const logout = async (req, res, next) => {
+  try {
+    await userLogoutSchema.validateAsync(req.body);
```

Comment:
To nie jest źle, tylko tak już propo wszystkich endpointów:
 - ograniczył bym się do walidacji joi w tych enpointach gdzie odczytujemy body, a jeśli już masz zamiar wymagać aby body było puste to JOI jest overkill'em
 ```
 if (Object.keys(req.body).length) {
   return res.status(400).......
}
 ```
 -  nie lubie jak używacie validateAsync. W joi jest funkcja validate która zwraca error
 ```
 const { error } = await userLogoutSchema.validate(req.body);
 if (error) {
   return res.status(400).......
 }
 ```
 w tym try/catchu może wystąpić więcej niż jeden błąd i jeśli chcemy rozróżnić je na 500 i 400 to musiał byś zrobić switch case i sprawdzać je po error.constructor i nie chcę wam tym mieszać ale w skrócie ta 400 nie zawsze powinna być 400